### ISSUE TITLE
Unexpected search results with index sorting

### ISSUE THREAD

**Issue by Saluev** [elastic/elasticsearch#36210]
**Elasticsearch version**: Version: 6.4.0, Build: default/tar/595516e/2018-08-17T23:18:47.308994Z

**Plugins installed**: []

**JVM version**: 10.0.2

**OS version**: macOS Mojave, Ubuntu 16.04.5 LTS

**Description of the problem including expected versus actual behavior**:

I created index with index sorting on two fields, `followersCnt` (long) and `created` (long). Then I filled it with documents; for a number of documents `followersCnt` field was not set. When running search with query
```
{
  "sort": [{"followersCnt": "desc"}],
  "size": 10
}
```
, I expect to get top 10 documents by `followersCnt`, but instead I get one top element and a number of random documents (on my production database) or no documents with `followersCnt` at all (locally after steps described below), i.e. all hit documents are those documents without `followersCnt` field.

**Steps to reproduce**:

```
// assuming Elasticsearch is on port 9202

curl -XPUT "http://localhost:9202/repro" -H 'Content-Type: application/json' -d'
{
    "settings" : {
        "index" : {
            "number_of_shards" : 5,
            "number_of_replicas" : 1,
            "sort" : {
               "order" : [
                  "desc",
                  "asc"
               ],
               "field" : [
                  "followersCnt",
                  "created"
               ]
            }
        }
    },
    "mappings" : {
       "doc" : {
          "properties" : {
             "name" : {
                "type" : "text"
             },
             "followersCnt" : {
                "type" : "long"
             },
             "created" : {
                "type" : "long"
             }
          }
       }
    }
}'

curl -XPUT "http://localhost:9202/repro/doc/0" -H 'Content-Type: application/json' -d'
{
  "name": "top user",
  "followersCnt": "1000000",
  "created": 100
}'

for i in {1..10}
do
  curl -XPUT "http://localhost:9202/repro/doc/$i" -H 'Content-Type: application/json' -d"
  {
   \"name\": \"cool user\",
   \"followersCnt\": \"$((1000 + $RANDOM))\",
   \"created\": $RANDOM
  }"
done

for i in {10001..11000}
do
  curl -XPUT "http://localhost:9202/repro/doc/$i" -H 'Content-Type: application/json' -d"
  {
   \"name\": \"lonely user\",
   \"created\": $RANDOM
 }"
done

curl -XPOST "http://localhost:9202/repro/_refresh"

curl -XPOST "http://localhost:9202/repro/_search" -H 'Content-Type: application/json' -d'
{
  "sort": [{"followersCnt": "desc"}],
  "size": 10
}' | json_pp | grep followersCnt

// expected output: sequence of numbers from 30000 to 10000
// real output: empty (none of the matched documents have `followersCnt` field)
```

**Comment by jimczi** [comment]
Unfortunately the bug is that documents are incorrectly sorted on disk when the natural order is reversed (`"order":"desc"`) and you have documents that don't have a value for the sort field. I opened https://issues.apache.org/jira/browse/LUCENE-8592 in Lucene to fix the sort for new indices but the workaround for new indices will certainly require a reindex with a version that has the fix. I'll update this issue with the outcome of the discussion in the Lucene issue.
Thanks a lot for reporting this issue @Saluev !

**Comment by jimczi** [comment]
The bug is fixed since Elasticsearch `6.6.0` where new indices are not affected. For old indices affected by this bug a full reindex is needed.
